src/Titon/Crypto/AbstractCipher.hh

```diff
@@ -0,0 +1,203 @@
+     */
+    protected string $mode;
+
+    /**
+     * The unique key.
+     *
+     * @var string
+     */
+    protected string $key;
+
+    /**
+     * Store the key and cipher method and verify its support.
+     *
+     * @param string $key
+     * @param string $method
+     */
+    public function __construct(string $key, string $method = self::AES_256_CBC) {
+        if (!static::isSupportedMethod($method)) {
+            throw new UnsupportedCipherException(sprintf('Unsupported cipher [%s] for [%s]', $method, static::class));
+
+        } else if (!static::isValidKey($key)) {
+            throw new InvalidKeyException(sprintf('Invalid key [%s] for [%s], must be 32 characters long', $key, static::class));
+        }
+
+        $this->key = $key;
+        $this->method = $method;
+        $this->algorithm = substr($method, 0, strlen($method) - 4);
+        $this->mode = substr($method, -3);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function decodePayload(string $payload): Payload {
+        $payload = json_decode(base64_decode($payload), true);
+
+        if (json_last_error() !== JSON_ERROR_NONE || !is_array($payload)) {
+            throw new InvalidPayloadException('Failed to decode payload');
+        }
+
+        $this->validatePayload($payload);
+        $this->validateMAC($payload);
+
+        return shape(
+            'iv' => $payload['iv'],
+            'data' => $payload['data'],
+            'mac' => $payload['mac']
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function encodePayload(string $data, string $iv): string {
+        return base64_encode(json_encode([
+            'iv' => $iv,
+            'data' => $data,
+            'mac' => $this->hashMAC($data, $iv)
+        ]));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getAlgorithm(): string {
+        return $this->algorithm;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getKey(): string {
+        return $this->key;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getMethod(): string {
+        return $this->method;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getMode(): string {
+        return $this->mode;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSupportedMethods(): Vector<string> {
+        return Vector {
+            self::AES_128_CBC,
+            self::AES_128_CFB,
+            self::AES_128_OFB,
+            self::AES_192_CBC,
+            self::AES_192_CFB,
+            self::AES_192_OFB,
+            self::AES_256_CBC,
+            self::AES_256_CFB,
+            self::AES_256_OFB,
+            self::BLOWFISH_CBC,
+            self::BLOWFISH_CFB,
+            self::BLOWFISH_OFB
+        };
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function isSupportedMethod(string $method): bool {
+        return in_array($method, static::getSupportedMethods());
```

Comment:
Since `getSupportedMethods` returns vector, should we use `get` and check null instead of `in_array`?